File: files/400-499/416/CH-416.R
library(tidyverse)
library(readxl)

path <- "400-499/416/CH-416 Number Series.xlsx"
input <- read_excel(path, range = "B3:B10")
test <- read_excel(path, range = "D3:D10")

fib_cache <- new.env(parent = emptyenv())

find_fib <- function(n) {
  key <- as.character(n)
  if (exists(key, envir = fib_cache, inherits = FALSE)) {
    return(get(key, envir = fib_cache, inherits = FALSE))
  }
  a <- 0
  b <- 1
  i <- 1

  while (b < n) {
    temp <- b
    b <- a + b
    a <- temp
    i <- i + 1
  }

  if (b == n) {
    value <- as.character(i)
  } else {
    value <- 'N/A'
  }

  assign(key, value, envir = fib_cache)
  value
}

result = input %>%
  mutate(n = map_chr(`Fibonacci Number`, find_fib))

all.equal(result$n, test$n)
# [1] TRUE

File: files/400-499/416/CH-416.py
import pandas as pd
import numpy as np
from functools import lru_cache

path = "400-499/416/CH-416 Number Series.xlsx"
input = pd.read_excel(path, usecols="B", nrows=8, skiprows=2)
test = pd.read_excel(path, usecols="D", nrows=8, skiprows=2)

@lru_cache(maxsize=None)
def is_fibonacci(n):
    a, b = 0, 1
    position = 0
    while a < n:
        a, b = b, a + b
        position += 1
    if a == n:
        return position
    else:
        return np.nan

input['Position'] = input['Fibonacci Number'].apply(is_fibonacci)

print(input["Position"].equals(test['n']))
# True